File: src/Java/myservicebus-rabbitmq/src/main/java/com/myservicebus/rabbitmq/RabbitMqPublishContext.java
package com.myservicebus.rabbitmq;

import com.myservicebus.PublishContext;
import com.myservicebus.tasks.CancellationToken;
import com.rabbitmq.client.AMQP;

/**
 * RabbitMQ-specific publish context exposing AMQP basic properties.
 */
public class RabbitMqPublishContext extends PublishContext {
    private final AMQP.BasicProperties.Builder properties;

    public RabbitMqPublishContext(Object message, CancellationToken cancellationToken) {
        super(message, cancellationToken);
        this.properties = new AMQP.BasicProperties.Builder().deliveryMode(2); // persistent
    }

    public AMQP.BasicProperties.Builder getProperties() {
        return properties;
    }
}

File: src/MyServiceBus.RabbitMq/RabbitMqPublishContext.cs
using System;
using System.Threading;
using MyServiceBus.Serialization;
using RabbitMQ.Client;

namespace MyServiceBus;

public class RabbitMqPublishContext : PublishContext
{
    public RabbitMqPublishContext(Type[] messageTypes, IMessageSerializer serializer, CancellationToken cancellationToken = default)
        : base(messageTypes, serializer, cancellationToken)
    {
        Properties = new BasicProperties
        {
            Persistent = true
        };
    }

    public BasicProperties Properties { get; }
}

public class RabbitMqPublishContextFactory : IPublishContextFactory
{
    public PublishContext Create(Type[] messageTypes, IMessageSerializer serializer, CancellationToken cancellationToken = default)
        => new RabbitMqPublishContext(messageTypes, serializer, cancellationToken);
}

